### PR TITLE
[Lib] Improve Windows diagnostics

### PR DESCRIPTION
**Pull request by BrianNixon**
Fixes all the bugs the compiler has been warning about, and several others I found while I was in there

### CODE REVIEW

File: lib/diagnostics.cpp [hunk 1]
```diff
@@ -496,7 +500,7 @@ int diagnostics_init(
         //   that for future use.
         lReturnValue = RegOpenKeyEx(
             HKEY_LOCAL_MACHINE,
-            _T("SOFTWARE\\Space Sciences Laboratory, U.C. Berkeley\\BOINC Setup"),
+            TEXT("SOFTWARE\\Space Sciences Laboratory, U.C. Berkeley\\BOINC Setup"),
```
Comment (AenBleidd) [inline lib/diagnostics.cpp:503]: `_T` correctly resolves to Unicode if needed.
`TEXT` by default is just a string literal as-is.
What is the reason of this change?

File: lib/diagnostics.cpp [hunk 2]
```diff
@@ -505,9 +509,9 @@ int diagnostics_init(
             // How large does our buffer need to be?
             dwSize = sizeof(boinc_install_dir);
 
-            lReturnValue = RegQueryValueEx(
+            lReturnValue = RegQueryValueExA(
```
Comment (AenBleidd) [inline lib/diagnostics.cpp:512]: Why changing from Unicode function to ANSI?

File: lib/diagnostics.cpp [hunk 3]
```diff
@@ -623,13 +627,8 @@ char* diagnostics_get_symstore() {
...
 //
-int diagnostics_set_symstore(char* project_symstore) {
-    if (!strlen(symstore)) {
-        int buffer_used = snprintf(symstore, sizeof(symstore), "%s", project_symstore);
-        if ((sizeof(symstore) == buffer_used) || (-1 == buffer_used)) {
-            symstore[sizeof(symstore)-1] = '\0';
-        }
-    }
+int diagnostics_set_symstore(const char* project_symstore) {
+    strlcpy(symstore, project_symstore, sizeof(symstore));
```
Comment (AenBleidd) [inline lib/diagnostics.cpp:631]: Please use `safe_strcpy` instead

File: lib/diagnostics_win.cpp [hunk 4]
```diff
@@ -303,9 +303,9 @@ int diagnostics_update_thread_list() {
                 pThreadEntry = diagnostics_find_thread_entry((DWORD)(uintptr_t)pThread->ClientId.UniqueThread);
 
                 if (pThreadEntry) {
-                    pThreadEntry->crash_kernel_time = (FLOAT)pThread->KernelTime.QuadPart;
-                    pThreadEntry->crash_user_time = (FLOAT)pThread->UserTime.QuadPart;
-                    pThreadEntry->crash_wait_time = (FLOAT)pThread->WaitTime;
+                    pThreadEntry->crash_kernel_time = pThread->KernelTime.QuadPart / 10000000.0;
```
Comment (AenBleidd) [inline lib/diagnostics_win.cpp:306]: `1e7` is shorter, I believe

**Review by AenBleidd — CHANGES_REQUESTED** [review]

File: lib/diagnostics.cpp (hunk 1 above)
Comment (BrianNixon) [reply lib/diagnostics.cpp:503]: API correctness
`TEXT` is the generic-text macro for the Windows API
`_T` is for the CRT

File: lib/diagnostics.cpp (hunk 2 above)
Comment (BrianNixon) [reply lib/diagnostics.cpp:512]: API correctness
`boinc_install_dir` is a `char` string
`RegQueryValueEx` is a generic-text function
The two should never be mixed

Comment (AenBleidd) [reply lib/diagnostics.cpp:512]: `RegQueryValueEx` is a macro (since `RegQueryValueExW` was added), and where it depends on the project settings what it will be resolved to, `RegQueryValueExA` or `RegQueryValueExW`.
Currently it resolves to `RegQueryValueExA`.
I suggest to keep it as is because in case we decide to make it wide char, it would be much easier to change to `boinc_install_dir` value on build error, rather than looking for all functions with `A` suffix at the end.

Comment (BrianNixon) [reply lib/diagnostics.cpp:512]: Changing to widechar is not going to be easy whichever way it’s done. In BOINC it could only reasonably be implemented in the context of ‘UTF-⁠8 everywhere’, in which case simply changing ANSI to Unicode and seeing what breaks will not even come close to getting it right. And the risk of that approach is that people will ‘fix’ the breakage by throwing in casts to make the compiler warnings go away while accusing Microsoft of not knowing what they’re doing… (Like [here](https://github.com/BOINC/boinc/blob/master/apps/concat.cpp#L95-L98 "apps/concat.cpp"), for instance.)

In theory, everything should _compile_ cleanly in both ANSI and Unicode. If it doesn’t, the code is using the API wrong. (Whether the application _works_ in both modes is an entirely different question…)

There are dozens of other examples of this inconsistency throughout BOINC; this instance only showed up as overlap between two concurrent patchsets which I haven’t kept in separate branches. The bigger question would be: are you interested in that kind of ‘code quality’ patch, or is it beyond the kind of thing that is worth spending time on?

Comment (AenBleidd) [reply lib/diagnostics.cpp:512]: In this particular case I prefer to keep it as-is, and do not change macro to the function with `A` suffix.
Thank you for understanding.

Answering you bigger question, it depends: I can't answer for all cases, each should be reviewed separately.

**Comment by davidpanderson** [comment]
General comments:
- we don't support 32-bit CPUs on Win anymore
- probably better to limit this PR to format codes

**Review by AenBleidd — APPROVED** [review]

**Comment by AenBleidd** [comment]
@BrianNixon, in general I'm ok with this PR. But next time please try to make PRs as atomic as possible.
Thank you for your contribution
